default/AI/ResearchAI.py

```diff
@@ -453,28 +492,45 @@ def generate_research_orders():
     rng = random.Random()
     rng.seed(fo.getEmpire().name + fo.getGalaxySetupData().seed)
 
-    research_reqs = calculate_research_requirements(empire)
+    calculate_research_requirements(empire)
     total_rp = empire.resourceProduction(fo.resourceType.research)
 
     if total_rp <= 0: # No RP available - no research.
         return
 
+    raw_priorities = {}
     priorities = {}
+    on_path_to = {}
     for tech_name in fo.techs():
-        priority = get_priority(rng, tech_name)
+        this_tech = fo.getTech(tech_name)
+        if not this_tech:
+            continue
+        raw_priorities[tech_name] = get_priority(rng, tech_name)
+    for tech_name, priority in raw_priorities.iteritems():
         if not tech_is_complete(tech_name) and priority >= 0:
-            turn_needed = float(research_reqs[tech_name][1]) / total_rp
-            priorities[tech_name] = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
+            turn_needed = max(research_reqs[tech_name][2], math.ceil(float(research_reqs[tech_name][1]) / total_rp))
+            adjusted_priority = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
```

Comment:
> These things is something we want to start early if we have the spare RP so we surely do not want to decrease the priority based on the time...

But if I have another tech with the same base priority and same cost, but taking half as long to finish (and no other differences at this level of abstraction), then I *do* want to prefer the one whose benefit I can get in half the time.  Even if we don't have enough RP to fully fund the faster one, it definitely won't take any longer than the slow one.

I acknowledge, however, that our current formula probably puts too great a weight on time to completion-- total cost should probably be the most important factor, and these time considerations merely secondary.

That was considering just the A vs B question.  I'll also take a moment to look at the average over a longer period of time.  I could do a series of techs with low min-time, or I could do a number of slower techs in parallel.  At the end of it I'll have accomplished the same total benefits over the same period of time.  But averages are not all that matter.  Each tech completion starts exerting it's benefit as soon as your get it, and so even with the same long term average rate, it's better to try to front load the benefits.